File: datasets/C++/Deepseek/DataStructures/hard5.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, m;
    cin >> n >> m;

    vector<vector<int>> adj(n+1);
    for (int i = 0; i < n-1; ++i) {
        int u, v;
        cin >> u >> v;
        adj[u].push_back(v);
        adj[v].push_back(u);
    }

    vector<vector<int>> distance(n+1, vector<int>(n+1, 0));
    for (int u = 1; u <= n; ++u) {
        vector<int> visited(n+1, -1);
        queue<int> q;
        q.push(u);
        visited[u] = 0;
        while (!q.empty()) {
            int current = q.front();
            q.pop();
            for (int neighbor : adj[current]) {
                if (visited[neighbor] == -1) {
                    visited[neighbor] = visited[current] + 1;
                    q.push(neighbor);
                }
            }
        }
        for (int v = 1; v <= n; ++v) {
            distance[u][v] = visited[v];
        }
    }

    vector<pair<int, int>> points(m);
    for (int i = 0; i < m; ++i) {
        cin >> points[i].first >> points[i].second;
    }

    int max_dist = 0;
    for (int i = 0; i < m; ++i) {
        int x1 = points[i].first;
        int y1 = points[i].second;
        for (int j = i+1; j < m; ++j) {
            int x2 = points[j].first;
            int y2 = points[j].second;
            int dist = distance[x1][x2] + distance[y1][y2];
            if (dist > max_dist) {
                max_dist = dist;
            }
        }
    }
    cout << max_dist << endl;

    return 0;
}

File: datasets/Python/Deepseek/DataStructures/hard5.py
import sys
from collections import deque

def main():
    input = sys.stdin
    n, m = map(int, input.readline().split())

    adj = [[] for _ in range(n+1)]
    for _ in range(n-1):
        u, v = map(int, input.readline().split())
        adj[u].append(v)
        adj[v].append(u)

    # Precompute distances from each node
    distance = [[0]*(n+1) for _ in range(n+1)]
    for u in range(1, n+1):
        visited = [-1]*(n+1)
        q = deque()
        q.append(u)
        visited[u] = 0
        while q:
            current = q.popleft()
            for neighbor in adj[current]:
                if visited[neighbor] == -1:
                    visited[neighbor] = visited[current] + 1
                    q.append(neighbor)
        for v in range(1, n+1):
            distance[u][v] = visited[v]

    points = []
    for _ in range(m):
        x, y = map(int, input.readline().split())
        points.append((x, y))

    max_dist = 0
    for i in range(m):
        x1, y1 = points[i]
        for j in range(i+1, m):
            x2, y2 = points[j]
            dist = distance[x1][x2] + distance[y1][y2]
            if dist > max_dist:
                max_dist = dist
    print(max_dist)

if __name__ == "__main__":
    main()